[3.x] Reduce texture memory consumption by reducing channels

mortarroad opened the pull request:
Reduce the channel count when saving, and store the expected format.
Enforce this format when loading.

Not all image formats support sRGB decoding, so we enforce using one of the supported formats (RGB8 and RGBA8, and some compressed formats) in GLES3.
This is the most robust way to implement this, because texture writing and loading is scattered across multiple places in the code.

If the texture needs sRGB decoding, we only reduce the channels to RGB8 or RGBA8 on import. (We could use other formats too, but at runtime it would be converted inside the GLES3 code.)

Also contains 2 extra fixes, which are required so that this works correctly.

Fixes https://github.com/godotengine/godot/issues/56523
Fixes https://github.com/godotengine/godot/issues/31759

mortarroad commented:
I couldn't measure any significant difference in performance.

However, I ran into an issue with sRGB handling:

In the preview when selecting the texture in the file browser LA8 is treated as sRGB, but L8 isn't. I have not been able to find the reason for this.
And in 3d with GLES3, neither L8 nor LA8 are treated as sRGB.
Setting the flag (`flags/srgb`) on import does not make any difference.

The only flag that does work is in the material itself (i.e. `flags_albedo_tex_force_srgb`).

I'm afraid we need to put this on hold until sRGB is properly implemented.

mortarroad commented:
Turns out standard OpenGL only has RGB and RGBA formats with sRGB support.
We should fall back onto those, if a texture has to be decoded as sRGB.

My GPU seems to support,
`GL_EXT_texture_sRGB`, which has L8 and LA8, and
`GL_EXT_texture_sRGB_R8`, which has R8,
but we cannot rely on that.

clayjohn commented:
This looks good from a quick review. But it is a risky change, so I'd like to review in further detail once I return from holidays.

mortarroad commented:
> This looks good from a quick review. But it is a risky change, so I'd like to review in further detail once I return from holidays.

Oh sorry, I think this needs further work before it can be merged.
The big issue is, that sRGB decoding is not supported for all texture formats. So it needs to be taken into account when storing a texture.

clayjohn commented:
> > This looks good from a quick review. But it is a risky change, so I'd like to review in further detail once I return from holidays.
> 
> Oh sorry, I think this needs further work before it can be merged. The big issue is, that sRGB decoding is not supported for all texture formats. So it needs to be taken into account when storing a texture.

Okay, sounds good. Let me know when you think it is ready and I will review it then!

mortarroad committed:
Fix detecting srgb textures

Prior, it was only detected if the format already supports srgb.

mortarroad committed:
Fix using the wrong image format in gles3

The image's format can be changed in `_get_gl_image_and_format()`, so we
have to use its returned value.

mortarroad committed:
Reduce texture memory consumption by reducing channels

Reduce the channel count when saving, and store the expected format.
Enforce this format when loading.

Not all image formats support sRGB decoding, so we enforce using one of
the supported formats (RGB8 and RGBA8, and some compressed formats) in
GLES3.
This is the most robust way to implement this, because texture writing
and loading is scattered across multiple places in the code.

If the texture needs sRGB decoding, we only reduce the channels to RGB8
or RGBA8 on import. (We could use other formats too, but at runtime it
would be converted inside the GLES3 code.)

Fixes https://github.com/godotengine/godot/issues/56523
Fixes https://github.com/godotengine/godot/issues/31759

mortarroad commented:
I've updated this.
I've fixed a few bugs in Godot's sRGB handling, which often resulted in not performing the sRGB transform. The channel optimization brought out these bugs more often.

This PR supersedes https://github.com/godotengine/godot/pull/52537

mortarroad commented:
@clayjohn This is ready for review, I think.

mortarroad commented:
Here's a little test project, that can be used to play around with the sRGB and compression settings:
[srgb_3.x.zip](https://github.com/godotengine/godot/files/9537263/srgb_3.x.zip)

It has one texture for each of L, LA, R, RG, RGB, and RGBA.
